File: Ruby/civitas/lib/civitasjuego.rb
#encoding: utf­8
require_relative 'gestor_estados'
require_relative 'jugador'
require_relative 'dado'
require_relative 'estados_juego'
require_relative 'tablero'
require_relative 'mazo_sorpresa'
require_relative 'casilla'
require_relative 'operaciones_juego'
require_relative 'casilla_calle'
require_relative 'casilla_sorpresa'
require_relative 'casilla_impuesto'
require_relative 'casilla_juez'
require_relative 'sorpresa_ir_a_carcel'
require_relative 'sorpresa_ir_a_casilla'
require_relative 'sorpresa_pagar_cobrar'
require_relative 'sorpresa_por_casa_hotel'
require_relative 'sorpresa_por_jugador'
require_relative 'sorpresa_salir_carcel'
require_relative 'sorpresa_calle'

module Civitas
  class Civitasjuego
    
    
    @titulo1 = TituloPropiedad.new("ETSIIT", 100, 1, 100, 500, 20)
    
    def initialize(_nombres)
      @gestor_estados = Gestor_estados.new()
      @gestor_estados.estado_inicial()
      @estado = Civitas::Estados_juego::INICIO_TURNO
      @jugadores=Array.new()
      for i in 0..(_nombres.size()-1)
        @jugadores << Jugador.new(_nombres.at(i))
      end
      @indice_jugador_actual= Dado.new.quien_empieza(@jugadores.size()-1)
      @tablero = Tablero.new(5)
      @mazo = MazoSorpresa.new(false)
      self.inicializar_tablero(@mazo)
      self.inicializar_mazosorpresas(@tablero)
    end
    
    attr_accessor :indice_jugador_actual, :estado, :gestor_estados, :tablero, :jugadores, :mazo
    
    
    
    def actualizar_info()
      @jugadores.at(self.indice_jugador_actual).to_sj()
      puts "La casilla actual es " + @jugadores.at(self.indice_jugador_actual).num_casilla_actual.to_s()
      if(@jugadores.at(self.indice_jugador_actual).es_bancarrota())
        _jugadores = Array.new()
        _jugadores = self.ranking()
        puts "El ranking es"
        for i in 0.._jugadores.size()
          puts _jugadores.at(i).to_sj()
        end
      end
    end
    
    def avanza_jugador()
      jugador_actual = @jugadores.at(@indice_jugador_actual)
      posicion_actual = jugador_actual.num_casilla_actual
      tirada = Dado.new.tirar()
      posicion_nueva = @tablero.nueva_posicion(posicion_actual, tirada)
      casilla =  @tablero.get_casilla(posicion_nueva)
      self.contabilizar_pasos_por_salida(jugador_actual)
      jugador_actual.mover_a_casilla(posicion_nueva)
      puts "Has caido en la casilla " + casilla.to_sc()
      casilla.recibe_jugador(@indice_jugador_actual, @jugadores)
      puts "Despues de recibe jugador"
      self.contabilizar_pasos_por_salida(jugador_actual)
    end
    
    def cancelar_hipoteca(_ip)
       
      return @jugadores.at(@indice_jugador_actual).cancelar_hipoteca(_ip)
    end
    
    def comprar()
      jugador_actual = @jugadores.at(@indice_jugador_actual)
      num_casilla_actual = jugador_actual.num_casilla_actual
      casilla = @tablero.get_casilla(num_casilla_actual)
      titulo=casilla.titulo
      res = jugador_actual.comprar(titulo)
      return res
    end
    
    def construir_casa(_ip)
      jugador_a = @jugadores.at(@indice_jugador_actual)
      return jugador_a.construir_casa(_ip)
    end
    
    def construir_hotel(_ip)
      jugador_a = @jugadores.at(@indice_jugador_actual)
      return jugador_a.construir_hotel(_ip)
    end
    
    def contabilizar_pasos_por_salida(_jugador)
      if(@tablero.porSalida()>0)
        _jugador.pasa_por_salida()
      end
    end
    
    def final_del_juego()
      fin = false
      for i in @jugadores
        if(i.es_bancarrota())
          fin = true
        end
      end
      return fin
    end
    
    def hipotecar(_ip)
      jugador_a = @jugadores.at(@indice_jugador_actual)
      return jugador_a.hipotecar(_ip)
    end
    
    def info_jugador_texto()
      jugador_a = @jugadores.at(@indice_jugador_actual)
      return jugador_a.to_sj()
    end
   
    def inicializar_mazosorpresas(_tablero)
      sorpresa1 = SorpresaPorCasaHotel.new(-120, "Te quito por cada casa y hotel 120 euros")
      sorpresa2 = SorpresaPorJugador.new(500, "Paga a los jugadores 500 euros")
      sorpresa3 = SorpresaPagarCobrar.new(1000, "Beneficio de 1000 eurazos")
      sorpresa4 = SorpresaIrACarcel.new(_tablero)
      sorpresa5 = SorpresaSalirCarcel.new(@mazo)
      sorpresa6 = SorpresaIrACasilla.new(_tablero, -150, "Vas a una casilla")
      sorpresa8 = SorpresaCalle.new(_tablero.get_casilla(1).titulo, @titulo1)
      @mazo.al_mazo(sorpresa8)
      @mazo.al_mazo(sorpresa8)
      @mazo.al_mazo(sorpresa8)
      @mazo.al_mazo(sorpresa8)
      @mazo.al_mazo(sorpresa8)
      @mazo.al_mazo(sorpresa8)
    end
    
    def inicializar_tablero(_mazo)
      titulo1 = TituloPropiedad.new("Granada", 100, 1, 100, 500, 20)
      titulo2 = TituloPropiedad.new("Tarragona", 120, 1, 150, 520, 25)
      titulo3 = TituloPropiedad.new("Melilla", 160, 2, 200, 540, 40)
      titulo4 = TituloPropiedad.new("Cadiz", 180, 2, 250, 560, 45)
      titulo5 = TituloPropiedad.new("Bilbao", 220, 2, 300, 580, 60)
      titulo6 = TituloPropiedad.new("Soria", 260, 3, 300, 600, 65)
      titulo7 = TituloPropiedad.new("Vigo", 280, 3, 400, 620, 70)
      titulo8 = TituloPropiedad.new("Baleares", 300, 3, 450, 640, 75)
      titulo9 = TituloPropiedad.new("Las Palmas", 350, 4, 500, 660, 85)
      titulo10 = TituloPropiedad.new("Asturias", 400, 4, 600, 680, 100)
      titulo11 = TituloPropiedad.new("Valencia", 450, 5, 700, 750, 200)
      titulo12 = TituloPropiedad.new("Guadalajara", 500, 5, 750, 800, 300)
      
      
     casilla1= CasillaCalle.new(titulo1)
     casilla2= CasillaSorpresa.new(_mazo, "Sorpresa 1: Nunca sabes lo que te va a tocar")
     casilla3= CasillaCalle.new(titulo2)
     casilla4= Casilla.new("Area de Descanso")
     casilla5= CasillaCalle.new(titulo3)
     casilla7= CasillaCalle.new(titulo4)
     casilla9= CasillaCalle.new(titulo5)
     casilla8= CasillaSorpresa.new(_mazo, "Sorpresa2: Ahi va una caja de comunidad")
     casilla10= CasillaCalle.new(titulo6)
     casilla11= CasillaCalle.new(titulo7)
     casilla6 = CasillaImpuesto.new(-250.0, "Control de alcoholemia")
     casilla12= CasillaCalle.new(titulo8)
     casilla13= CasillaCalle.new(titulo9)
     casilla15= CasillaCalle.new(titulo10)
     casilla14= CasillaSorpresa.new(_mazo, "Sorpresa3")
     casilla16= CasillaCalle.new(titulo11)
     casilla17 = CasillaCalle.new(titulo12)

      @tablero.aniade_casilla(casilla1)
      @tablero.aniade_casilla(casilla3)
      @tablero.aniade_casilla(casilla2)
      @tablero.aniade_casilla(casilla8)
      @tablero.aniade_casilla(casilla5)
      @tablero.aniade_casilla(casilla6)
      @tablero.aniade_casilla(casilla7)
      @tablero.aniade_casilla(casilla4)
      @tablero.aniade_casilla(casilla9)
      @tablero.aniade_juez()
      @tablero.aniade_casilla(casilla10)
      @tablero.aniade_casilla(casilla11)
      @tablero.aniade_casilla(casilla12)
      @tablero.aniade_casilla(casilla13)
      @tablero.aniade_casilla(casilla14)
      @tablero.aniade_casilla(casilla15)
      @tablero.aniade_casilla(casilla16)
      @tablero.aniade_casilla(casilla17)
    
    end
    
    def pasar_turno()
      puts "Pasar turno primero vale " + @indice_jugador_actual.to_s()
      @indice_jugador_actual = (@indice_jugador_actual+1)%@jugadores.length
      puts "Pasar turno despues vale " + @indice_jugador_actual.to_s()
    end
    
    def ranking()
      puts "estoy en ranking"
      _jugadores=Array.new()
      for i in 0..(@jugadores.length()-1)
        _jugadores << @jugadores.at(i)
      end
      for i in 0..(_jugadores.length()-2)
        for j in i+1..(_jugadores.length()-1)
          if(_jugadores.at(i).compare_to(_jugadores.at(j)))
            aux = Jugador.new(_jugadores.at(i))
            _jugadores.insert(i, _jugadores.at(j))
            _jugadores.insert(j, aux)
          end
        end
      end
      for i in 0..(_jugadores.length-1)
        puts _jugadores.at(i).to_sj()
      end
    end
    
    def salir_carcel_pagando()
      jugador_a = @jugadores.at(@indice_jugador_actual)
      return jugador_a.salir_carcel_pagando()
    end
    
    def salir_carcel_tirando()
      jugador_a = @jugadores.at(@indice_jugador_actual)
      return jugador_a.salir_carcel_tirando()
    end
    
    def siguiente_paso()
      jugador_actual = @jugadores.at(@indice_jugador_actual)
      operacion = @gestor_estados.operaciones_permitidas(jugador_actual, @estado)
      if(operacion == Civitas::Operaciones_juego::PASAR_TURNO)
        self.pasar_turno()
        self.siguiente_paso_completado(operacion)
      else
        if(operacion == Civitas::Operaciones_juego::AVANZAR)
          self.avanza_jugador()
          self.siguiente_paso_completado(operacion)
        end
      end
      return operacion
    end
    
    def siguiente_paso_completado(_operacion)
      @estado = @gestor_estados.siguiente_estado(@jugadores.at(@indice_jugador_actual), @estado, _operacion)
    end
    
    def vender(_ip)
      jugador_a= @jugadores.at(@indice_jugador_actual)
      return jugador_a.vender(_ip)
    end
      
    def get_casilla_actual()
      indice = @jugadores.at(@indice_jugador_actual).num_casilla_actual
      return @tablero.get_casilla(indice)
    end
    
    def mostar_tan()
      for i in 0..(@tablero.tamanio()-1)
        puts "El tablero " + @tablero.casillas.at(i).to_sc()
      end
    end
      
    
    #@author JuanCarlos Gonzalez Quesada
    #EXAMEN
    def robar_propiedad()
      jugador_actual = @jugadores.at(@indice_jugador_actual)
      casilla = @tablero.get_casilla(jugador_actual.numCasillaActual)
      titulo_robado = casilla.get_titulo_propiedad_a_robar()
      jugador_actual.robar_propiedad(titulo_robado)
    end
    
  end
end

File: Java/civitas/src/civitas/CivitasJuego.java
package civitas;

import GUI.Dado;
import java.lang.reflect.Array;
import java.util.ArrayList;
import java.util.Collections;

public class CivitasJuego {
    private int indiceJugadorActual;
    private EstadosJuego estado;
    private GestorEstados gestorEstados;
    private Tablero tablero;
    private ArrayList<Jugador> jugadores;
    private MazoSorpresa mazo;
    
    private TituloPropiedad titulo1 = new TituloPropiedad("Despacho de Nuria", (float) 100, 1, 100, 500, 20);
    private TituloPropiedad titulo2 = new TituloPropiedad("Despacho de Kawtar", (float) 120, 1, 150, 520, 25);
    
    
    public CivitasJuego(ArrayList<String> _nombres){
        jugadores = new ArrayList<>();
        for(int i=0; i<_nombres.size(); i++){
            Jugador _jugador = new Jugador(_nombres.get(i));
            jugadores.add(_jugador);
        }
        gestorEstados = new GestorEstados();
        mazo = new MazoSorpresa();
        tablero = new Tablero(5);
        this.inicializarTablero(mazo);
        this.inicializarMazoSorpresas(tablero);
        estado = gestorEstados.estadoInicial();;
        this.indiceJugadorActual = Dado.getInstance().quienEmpieza(jugadores.size());
        
    }
    
    void actualizarInfo(){
        jugadores.get(this.indiceJugadorActual).toString();
        if(jugadores.get(this.indiceJugadorActual).enBancarrota()){
            ArrayList<Jugador> _jugadores = new ArrayList<>();
            _jugadores = this.ranking();
            System.out.println("El Ranking es ");
            for(int i=0; i<_jugadores.size(); i++){
                System.out.println(_jugadores.get(i).toString());
            }
        }
    }
    private void avanzaJugador(){
        Jugador jugadorActual = this.getJugadorActual();
        int posicionActual = jugadorActual.getNumCasillaActual();
        int tirada = Dado.getInstance().tirar();
        int posicionNueva = tablero.nuevaPosicion(posicionActual, tirada);
        Casilla casilla = tablero.getCasilla(posicionNueva);
        this.contabilizarPasosPorSalida(jugadorActual);
        jugadorActual.moverACasilla(posicionNueva);
        casilla.recibeJugador(indiceJugadorActual, jugadores);
        Diario.getInstance().ocurreEvento("La casilla es " + (this.tablero.getCasilla(jugadorActual.getNumCasillaActual())).toString());
        this.contabilizarPasosPorSalida(jugadorActual);
    }
    
    public boolean cancelarHipoteca(int _ip){
        Jugador jugadorA = this.jugadores.get(indiceJugadorActual);
        return jugadorA.cancelarHipoteca(_ip);
    }
    
    public boolean comprar(){
        Jugador jugadorActual=jugadores.get(indiceJugadorActual);
        int numCasillaActual = jugadorActual.getNumCasillaActual();
        Casilla casilla = tablero.getCasilla(numCasillaActual);
        TituloPropiedad titulo =((CasillaCalle) casilla).getTituloPropiedad();
        boolean res = jugadorActual.comprar(titulo);
        return res;
    }
    
    public boolean construirCasa(int _ip){
        Jugador jugadorA = this.jugadores.get(indiceJugadorActual);
        return jugadorA.construirCasa(_ip);
    }
    
    public boolean construirHotel(int _ip){
        Jugador jugadorA = this.jugadores.get(indiceJugadorActual);
        return jugadorA.construirHotel(_ip);
    }
    
    private void contabilizarPasosPorSalida(Jugador _jugadorActual){
        if(tablero.getPorSalida()>0)
            _jugadorActual.pasaPorSalida();
    }
    
    public boolean finalDelJuego(){
        boolean fin=false;
        for(Jugador j: jugadores){
            if(j.enBancarrota())
                fin=true;
        }
        return fin;
    }
    
    public Casilla getCasillaActual(){
        int indice = jugadores.get(indiceJugadorActual).getNumCasillaActual();
        return tablero.getCasilla(indice);
    }
    
    public Jugador getJugadorActual(){
        return jugadores.get(indiceJugadorActual);
    }
    
    public boolean hipotecar(int _ip){
        Jugador jugadorA = this.jugadores.get(indiceJugadorActual);
        return jugadorA.hipotecar(_ip);
    }
    
    public String infoJugadorTexto(){
        Jugador jugadorA = this.jugadores.get(indiceJugadorActual);
        return jugadorA.toString();
    }
    
    public void inicializarMazoSorpresas(Tablero _tablero){
        Sorpresa sorpresa1 = new SorpresaPorCasaHotel(-120, "Tienes que pagar 120 euros para arreglar el megáfono de Mancia");
        Sorpresa sorpresa2 = new SorpresaIraCarcel(_tablero);
        Sorpresa sorpresa3 = new SorpresaIraCasilla(_tablero, -150, "Pagas 150 y vas a una casilla");
        Sorpresa sorpresa4 = new SorpresaSalirCarcel(this.mazo);
        Sorpresa sorpresa5 = new SorpresaPorJugador(500, "Te ha pilado copiándote Nacho (DDSI) paga a tus compañeros para que no te delanten");
        Sorpresa sorpresa6 = new SorpresaPagarCobrar(1000, "Has ganado un pulso a Extremera, recibe 1000 euros de un Proxeneta");
        Sorpresa sorpresa7 = new SorpresaEspeculador(1000);
        
        Sorpresa sorpresa8 = new SorpresaConvierteCalle(1,"Convertir y robar", ((CasillaCalle)_tablero.getCasilla(1)).getTituloPropiedad() , titulo1, _tablero);
        Sorpresa sorpresa9 = new SorpresaConvierteCalle(1,"Convertir y robar", ((CasillaCalle)_tablero.getCasilla(8)).getTituloPropiedad() , titulo2, _tablero);
        mazo.alMazo(sorpresa8);
        mazo.alMazo(sorpresa9);
        mazo.alMazo(sorpresa8);
        mazo.alMazo(sorpresa9);
        mazo.alMazo(sorpresa8);
        mazo.alMazo(sorpresa9);
        mazo.alMazo(sorpresa8);
    }
    
    public void inicializarTablero(MazoSorpresa _mazoSorpresas){
        TituloPropiedad titulo1 = new TituloPropiedad("Despacho de Lastra", (float) 100, 1, 100, 500, 20);
        TituloPropiedad titulo2 = new TituloPropiedad("Despacho de Cabrerizo", (float) 120, 1, 150, 520, 25);
        TituloPropiedad titulo3 = new TituloPropiedad("Despacho de Baldomero", (float) 160, 2, 200, 540, 40);
        TituloPropiedad titulo4 = new TituloPropiedad("Despacho de Capel", (float) 180, 2, 250, 560, 45);
        TituloPropiedad titulo5 = new TituloPropiedad("Despacho de Oresti", (float) 220, 2, 300, 580, 60);
        TituloPropiedad titulo6 = new TituloPropiedad("Despacho de Anguita", (float) 260, 3, 300, 600, 65);
        TituloPropiedad titulo7 = new TituloPropiedad("Despacho de Melero", (float) 280, 3, 400, 620, 70);
        TituloPropiedad titulo8 = new TituloPropiedad("Despacho de La'Patri", (float) 300, 3, 450, 640, 75);
        TituloPropiedad titulo9 = new TituloPropiedad("Despacho de Extremera", (float) 350, 4, 500, 660, 85);
        TituloPropiedad titulo10 = new TituloPropiedad("Despacho de Sevilla", (float) 400, 4, 600, 680, 100);
        TituloPropiedad titulo11 = new TituloPropiedad("Despacho de Urbano", (float) 450, 5, 700, 750, 200);
        TituloPropiedad titulo12 = new TituloPropiedad("Despacho de Mancia", (float) 500, 5, 750, 800, 300);
                
        Casilla casilla1 = new CasillaCalle(titulo1);
        Casilla casilla2 = new CasillaSorpresa(_mazoSorpresas, "Examen Sorpresa de Manuel Capel: PON EL CÓDIGO: ExamenUbu16");    
        Casilla casilla3 = new CasillaCalle(titulo2);     
        Casilla casilla4 = new Casilla("Has entrado en el despacho de Lozano. Tómate un descanso");    
        Casilla casilla5 = new CasillaCalle(titulo3);
        Casilla casilla6 = new CasillaCalle(titulo4);
        Casilla casilla7 = new CasillaCalle(titulo5);
        Casilla casilla8 = new CasillaSorpresa(_mazoSorpresas, "Te has topado con Pedro Smith, por guardar su maletin te da una SORPRESA");
        Casilla casilla9 = new CasillaCalle(titulo6);
        Casilla casilla10 = new CasillaCalle(titulo7);
        Casilla casilla11 = new CasillaImpuesto( (int) -120.0, "Control de alcoholemia. Sevilla y tú habeís sido parados por la poli despues de la fiesta");
        Casilla casilla12 = new CasillaCalle(titulo8);
        Casilla casilla13 = new CasillaCalle(titulo9);
        Casilla casilla14 = new CasillaCalle(titulo10);
        Casilla casilla15 = new CasillaSorpresa(_mazoSorpresas, "SORPRESA! Te la las pácticas Nawja! Coge una sorpesa");
        Casilla casilla16 = new CasillaCalle(titulo11);
        Casilla casilla17 = new CasillaCalle(titulo12);
        
        tablero.aniadeCasilla(casilla1);
        tablero.aniadeCasilla(casilla3);
        tablero.aniadeCasilla(casilla2);
        tablero.aniadeCasilla(casilla8);
        tablero.aniadeCasilla(casilla4);
        tablero.aniadeCasilla(casilla5);
        tablero.aniadeCasilla(casilla6);
        tablero.aniadeCasilla(casilla7);
        tablero.aniadeCasilla(casilla9);
        tablero.aniadeJuez();
        tablero.aniadeCasilla(casilla10);
        tablero.aniadeCasilla(casilla11);
        tablero.aniadeCasilla(casilla12);
        tablero.aniadeCasilla(casilla13);
        tablero.aniadeCasilla(casilla14);
        tablero.aniadeCasilla(casilla15);
        tablero.aniadeCasilla(casilla16);
        tablero.aniadeCasilla(casilla17);
    }
    
    public void pasarTurno(){
        this.indiceJugadorActual=(this.indiceJugadorActual+1)%this.jugadores.size();
    }
    
    public ArrayList<Jugador> ranking(){
        ArrayList<Jugador> _jugadores = new ArrayList<>();
        Collections.sort(jugadores);
        _jugadores=jugadores;
        return _jugadores;
    }
    
    public boolean salirCarcelPagando(){
        Jugador jugadorA = this.jugadores.get(indiceJugadorActual);
        return jugadorA.salirCarcelPagando();
    }
    
    public boolean salirCarcelTirando(){
        Jugador jugadorA = this.jugadores.get(indiceJugadorActual);
        return jugadorA.salirCarcelTirando();
    }
    
    public OperacionesJuego siguientePaso(){
        Jugador jugadorActual = jugadores.get(indiceJugadorActual);
        OperacionesJuego operacion = gestorEstados.operacionesPermitidas(jugadorActual, estado);
        if(operacion == OperacionesJuego.PASAR_TURNO){
            pasarTurno();
            siguientePasoCompletado(operacion);
        }
        else 
            if(operacion == OperacionesJuego.AVANZAR){
                avanzaJugador();
                siguientePasoCompletado(operacion);
        }
        return operacion;
    }
    
    public void siguientePasoCompletado(OperacionesJuego operacion){
        estado = gestorEstados.siguienteEstado(jugadores.get(indiceJugadorActual), estado, operacion);
    }
    
    public boolean vender(int _ip){
        Jugador jugadorA = this.jugadores.get(indiceJugadorActual);
        return jugadorA.vender(_ip);
    } 
    
    
    
    
    //EXAMEN
    public boolean robarPropiedad(){
        System.out.println("Estoy en robarPropiedad de Civitas");
        Jugador jugadorActual = jugadores.get(this.indiceJugadorActual);
        Casilla casilla = tablero.getCasilla(jugadorActual.getNumCasillaActual());
        TituloPropiedad tituloARobar = ((SorpresaCalle) casilla).getTituloPropiedadARobar();
        jugadorActual.robarPropiedad(tituloARobar);
        return true;
    }
}
